### PR TITLE
Show filtered records in Query Records Display

### PR DESCRIPTION
**Pull request by joiecosby**
This PR will resolve https://github.com/jetstreamapp/jetstream/issues/45

I updated the display of records section in SalesforceRecordDataTable to now account for filtering. This can be seen in the screenshot below. I know the design for this was not finalized but I decided to export a new const where we can hold different values related to DataTable. The goal would be that we could easily add to it in the future.

Screen Shot 2023-04-22 at 20 56 04

### CODE REVIEW

File: libs/ui/src/lib/data-table/DataTable.tsx [hunk 1]
```diff
@@ -176,6 +176,10 @@ export interface DataTableProps<T = RowWithKey, TContext = Record<string, any>>
   onSortedAndFilteredRowsChange?: (rows: readonly T[]) => void;
 }
 
+export const DataTableProperties = {
```
Comment (paustint) [inline libs/ui/src/lib/data-table/DataTable.tsx]: @joiecosby - This variable is global which means it can apply to any and every table (at the same time) and if there happen to be two tables displayed on the page at the same time, then the  value would get modified by the second table, which could cause the data to display incorrect data.

The `<DataTable />` component already emits `onSortedAndFilteredRowsChange` so any component implementing a the DataTable has access to the filtered data.

I think a cleaner approach would be something like in `SalesforceRecordDataTable` -> `handleSortedAndFilteredRowsChange`, store `rows.length` as a state variable

e.x.
```
const [visibleRecordCount, setVisibleRecordCount] = useState<number>(0);

/// in  handleSortedAndFilteredRowsChange
setVisibleRecordCount(rows.length);
```

In summary -> 
* we don't want to store global state in most cases, especially not for things that are related to a component's rendering cycle
* We rarely want to break react's standard data flow (parent passes data to children through props, and children pass data back to parents as events (e.x. call function that parent passed to child as a prop))

**Review by paustint — CHANGES_REQUESTED** [review]

**Comment by paustint** [comment]
@joiecosby - Is the PR name for this one accurate? Seems like the name suggests it is for a different ticket.

File: libs/ui/src/lib/data-table/DataTable.tsx (hunk 1 above)
Comment (joiecosby) [reply libs/ui/src/lib/data-table/DataTable.tsx]: @paustint so does the useCallback() hook help make the data in DataTable responsive and update to date? Is that handling the vent passed back?

```
    const handleSortedAndFilteredRowsChange = useCallback(
      (rows: RowSalesforceRecordWithKey[]) => {
        onFilteredRowsChanged(rows.map(({ _record }) => _record));

        setVisibleRecordCount(rows.length);
      },
      [onFilteredRowsChanged]
    );
```

File: libs/ui/src/lib/data-table/SalesforceRecordDataTable.tsx [hunk 2]
```diff
@@ -99,6 +99,7 @@ export const SalesforceRecordDataTable: FunctionComponent<SalesforceRecordDataTa
     const [nextRecordsUrl, setNextRecordsUrl] = useState<Maybe<string>>(null);
     const [globalFilter, setGlobalFilter] = useState<string | null>(null);
     const [selectedRows, setSelectedRows] = useState<ReadonlySet<string>>(() => new Set());
+    const [visibleRecordCount, setVisibleRecordCount] = useState<number>(0);
```
Comment (paustint) [inline libs/ui/src/lib/data-table/SalesforceRecordDataTable.tsx]: @joiecosby - usually it is best if you can try to initialize any state values with an accurate value if possible, and I think here we can. Reason being is that it ensures that if the UI renders initially it will show the correct value instead of showing an incorrect value momentarily then showing the correct value.

Also, the type can be inferred by the value so you generally don't need to include a generic type (`<number>`) unless the type cannot be inferred.

```suggestion
    const [visibleRecordCount, setVisibleRecordCount] = useState(records.length);
```

Also, you might notice that many places use the function syntax `useState(() => records.length);` -> this syntax should be used if deriving the value is expensive (anything that is not a simple value being used) because the function syntax gets called lazily, where the normal syntax gets called on every render no matter what even if there is already a value stored in state.

**Review by paustint — CHANGES_REQUESTED** [review]

File: libs/ui/src/lib/data-table/SalesforceRecordDataTable.tsx (hunk 2 above)
Comment (paustint) [reply libs/ui/src/lib/data-table/SalesforceRecordDataTable.tsx]: @joiecosby - Here is an example that I just came across. State was initialized as empty array then an useEffect was used to set the data, which happens on a subsequent render.

In this case, the computation was sort of expensive (loop through a bunch of items) so in addition to setting it on init with a function (lazy init), I also used a custom hook in the codebase called `useNonInitialEffect` - it just means it does not run on first render, but runs on ever change after that.

I also extracted to a function, outside the component, since there was a lot of code. The reason the function was moved outside the component was because that means it does not have to get re-created on every render. It also means the function must take in all the data as a parameter instead of having direct access to variables within the component's scope.

Here is a demo of the UI experience before and after:
BEFORE: https://jam.dev/c/f8cf3c05-1a10-4e37-b0e7-90887458f8b5
AFTER: https://jam.dev/c/b39fb3fa-7121-45c5-8ace-bea276d3ceee

image

File: libs/ui/src/lib/data-table/DataTable.tsx (hunk 1 above)
Comment (paustint) [reply libs/ui/src/lib/data-table/DataTable.tsx]: @joiecosby - the usecallback just means that the function does not get re-created on every render, and only gets re-created if `onFilteredRowsChanged` changes (that is in the dependency array). The reason why this is done is because this function is passed down to child components and we want to limit re-renders of children.
(some children may need to call this function in a useEffect, which means this function would be a dependency of those useEffect hooks).

This is a SUPER long article, but I highly suggest taking the time to read and digest it: https://overreacted.io/a-complete-guide-to-useeffect/

File: libs/ui/src/lib/data-table/SalesforceRecordDataTable.tsx (hunk 2 above)
Comment (joiecosby) [reply libs/ui/src/lib/data-table/SalesforceRecordDataTable.tsx]: That is a great visual representation of what you are trying to explain. So because the logic for setting visibleRecordCount is straightforward and not costly, we can set it by just using `useState(records.length);` 

Is the logic below, another good example of what you are trying to get across? That we init at rendering by calling getModalTitle() and then using useNonInitialEffect() to render any changes more forward after the initial init. 

Quick question, is useNonInitialEffect() tracking action in a similar fashion to the point you made in [this](https://github.com/jetstreamapp/jetstream/pull/324#discussion_r1176671178) comment? I believe so but want to double check! I can look at the article you mentioned too.

```
  const [modalTitle, setModalTitle] = useState(() => getModalTitle(action));

  const [downloadModalOpen, setDownloadModalOpen] = useState(false);

  useNonInitialEffect(() => {
    setModalTitle(getModalTitle(action));
  }, [action]);
```

Comment (joiecosby) [reply libs/ui/src/lib/data-table/SalesforceRecordDataTable.tsx]: Two more questions for you:

1. since records is getting set in a const above and is getting set to `any[] | undefined` during the init do we ever have to worry about setting the const like below? My assumption is yes just in case records is ever undefined.

` const [visibleRecordCount, setVisibleRecordCount] = useState(records?.length);`

2. If records isn't being set until the useEffect is being called, would visibleRecordCount technically still be 0 on it?

Comment (paustint) [reply libs/ui/src/lib/data-table/SalesforceRecordDataTable.tsx]: `useNonInitialEffect` - this is just a custom thing I built (or probably found on a blog article and slightly modified) - `useEffect` runs whenever dependencies change and cause a re-render, with `useNonInitialEffect` is the exact same except won't run on the very first render.
This is sort of like a micro-optimization, but sometimes - mostly in very complex components, it does make a difference.

> do we ever have to worry about setting the const like below
If `records` can be undefined in the props type definition, then we would absolutely need to do `useState(records?.length);`  - which as a result, `visibleRecordCount` would have the type `number | undefined`.
Many times it is nice to ensure a value has a value, in this case `useState(records?.length || 0);` would be a good solution.

> If records isn't being set until the useEffect is being called, would visibleRecordCount technically still be 0 on it?
Yes - BUT there is a caveat. Once records is defined, it shows up in the ui right away THEN the useEffect is called on a second render - so there is still room for a jumpy UI. Usually the jumpy UI only happens on the initial render unless there is a lot of dependent updates being made that result in spanning multiple renders.

You will see at some places around the codebase `useReducer` - this is a solution to that problem above, where instead of having a bunch of `useEffect` all updating state based on prior useEffect, all the updates can be made at once. (useReducer isn't used very often and is only really only helpful in fairly complex situations)

**Review by paustint — APPROVED** [review]
